File: NodeJS/src/Snake.js
class Snake {
    #snake
    #food
    #dir
    #prev_dir
    #width
    #height
    #running
    constructor(width, height) {
        this.#width = width;
        this.#height = height;
        this.#snake = [[parseInt(height / 2), parseInt(width / 2)], [parseInt(height / 2), parseInt(width / 2)]];
        this.#dir = [0, 1];
        this.#prev_dir = this.#dir;
        this.#food = null;
        this.#running = true;
        this.spawnFood();
    }
    update() {
        if (!this.#running) return false;
        let newHead = [
            (this.getHeight() + this.#dir[0] + this.#snake[0][0]) % this.getHeight(),
            (this.getWidth() + this.#dir[1] + this.#snake[0][1]) % this.getWidth()
        ];
        this.#prev_dir = this.#dir;
        if (!this.onFood(newHead))
            this.#snake.pop();
        if (this.onSnake(newHead)) {
            this.#running = false;
            return false;
        }
        this.#snake.unshift(newHead);
        if (this.onFood(newHead))
            this.spawnFood();
        return true;
    }
    changeDir(dir) {
        if (!this.#running || dir[0] === this.#prev_dir[0] || dir[1] === this.#prev_dir[1])
            return;
        this.#dir = dir;
    }
    onSnake([x, y]) {
        let ans = false;
        this.#snake.forEach(cell => {
            ans = ans || (x === cell[0] && y === cell[1]);
        });
        return ans;
    }
    onFood([x, y]) {
        return this.#food && (x === this.#food[0] && y === this.#food[1]);
    }
    spawnFood() {
        if (!this.#running) return;
        let free = [];
        for (let x = 0; x < this.getHeight(); x++)
            for (let y = 0; y < this.getWidth(); y++) {
                if (!this.onSnake([x, y]))
                    free.push([x, y]);
            }
        this.#food = free[Math.floor(Math.random() * free.length)];
    }
    getScore() { return 50 * (this.#snake.length - 2); }
    getWidth() { return this.#width; }
    getHeight() { return this.#height; }
    isRunning() { return this.#running; }
}

module.exports = Snake;

File: Go/snake.go
package main

import (
	"math/rand"
	"time"
)

type GameState int
const (
	Running GameState = iota
	Paused
	GameOver
	Win
)

type Game struct {
	snake	[][2]int
	food	[2]int
	dir		[2]int
	pDir	[2]int
	width	int
	height	int
	state	GameState
}

func newGame(w, h int) *Game {
	rand.Seed(time.Now().UnixNano())
	game := Game{width: w, height: h}
	game.init()
	return &game
}

func (g *Game) init() {
	g.snake = [][2]int{{g.width / 2, g.height / 2}}
	g.food = [2]int{g.width / 2 + 1, g.height / 2}
	g.dir = [2]int{1, 0}
	g.pDir = g.dir
	g.state = Running
}

func (g *Game) changeDir(x, y int) {
	if g.state != Running { return }
	if !(x + g.dir[0] == 0 && y + g.dir[1] == 0) {
		g.pDir = [2]int{x, y}
	}
}

func (g *Game) spawn() bool {
	emptyCells := make([][2]int, 0, g.width * g.height)
	for i := 0; i < g.width; i++ {
		for j := 0; j < g.height; j++ {
			if !g.onSnake(i, j) && !g.onFood(i, j) {
				emptyCells = append(emptyCells, [2]int{i, j})
			}
		}
	}
	if len(emptyCells) == 0 {
		g.food = [2]int{g.width + 1, g.height + 1}
		return false
	}
	g.food = emptyCells[int(rand.Intn(len(emptyCells)))]
	return true
}

func (g *Game) update() GameState {
	if g.state != Running { return g.state }
	g.dir = g.pDir
	newHead := [2]int{
		(g.snake[len(g.snake) - 1][0] + g.dir[0] + g.width) % g.width,
		(g.snake[len(g.snake) - 1][1] + g.dir[1] + g.height) % g.height,
	}
	if g.onFood(newHead[0], newHead[1]) {
		if !g.spawn() {
			g.state = Win
		}
	} else {
		g.snake = g.snake[1:]
	}
	if g.onSnake(newHead[0], newHead[1]) {
		g.state = GameOver
	}
	g.snake = append(g.snake, newHead)
	return g.state
}

func (g *Game) onSnake(x, y int) bool {
	for _, cell := range g.snake {
		if cell[0] == x && cell[1] == y {
			return true
		}
	}
	return false
}

func (g *Game) onFood(x, y int) bool {
	return (x == g.food[0] && y == g.food[1])
}

func (g *Game) getScore() int {
	return (len(g.snake) - 2) * 50
}

func (g *Game) togglePause() GameState {
	if g.state == Running {
		g.state = Paused
	} else if g.state == Paused {
		g.state = Running
	}
	return g.state
}
